### PR TITLE
Define `fpp_t` and `f_cnt_t` to be of `size_t`

### PR DESCRIPTION
**Pull request by sakertooth**
Defines  `fpp_t` and `f_cnt_t` to be of `size_t`. Most of the codebase used `fpp_t` to represent the size of an array of sample frames, which is incorrect, given that `fpp_t` was only 16 bits when sizes greater than `32767` are very possible.

There were discussions about using `std::size_t`, but I left it as `size_t` so that it can be adopted a bit easier in the code, and not end up in a situation where we mix up `std::size_t` and `size_t`.

### CODE REVIEW

File: include/AudioDevice.h [hunk 1]
```diff
@@ -93,21 +93,21 @@ class AudioDevice
...
 	int convertToS16(const SampleFrame* _ab,
-						const fpp_t _frames,
+						const size_t _frames,
 						int_sample_t * _output_buffer,
 						const bool _convert_endian = false );
 
 	// clear given signed-int-16-buffer
 	void clearS16Buffer( int_sample_t * _outbuf,
-							const fpp_t _frames );
+							const size_t _frames );
```
Comment (Rossmaxx) [inline include/AudioDevice.h:110]:
```suggestion
							const size_t _frames);
```
Style

File: include/AudioEngine.h [hunk 2]
```diff
@@ -270,14 +270,14 @@ class LMMS_EXPORT AudioEngine : public QObject
 		return m_fifoWriter != nullptr;
 	}
 
-	void pushInputFrames( SampleFrame* _ab, const f_cnt_t _frames );
+	void pushInputFrames( SampleFrame* _ab, const size_t _frames );
```
Comment (Rossmaxx) [inline include/AudioEngine.h:273]: Style here too.

File: include/AudioEngineProfiler.h [hunk 3]
```diff
@@ -46,7 +46,7 @@ class AudioEngineProfiler
 		m_periodTimer.reset();
 	}
 
-	void finishPeriod( sample_rate_t sampleRate, fpp_t framesPerPeriod );
+	void finishPeriod( sample_rate_t sampleRate, size_t framesPerPeriod );
```
Comment (Rossmaxx) [inline include/AudioEngineProfiler.h:49]: Here too.

File: include/AudioDevice.h [hunk 4]
```diff
@@ -93,21 +93,21 @@ class AudioDevice
 protected:
 	// subclasses can re-implement this for being used in conjunction with
 	// processNextBuffer()
-	virtual void writeBuffer(const SampleFrame* /* _buf*/, const fpp_t /*_frames*/) {}
```
Comment (Rossmaxx) [inline include/AudioDevice.h:96]: Would it make sense to remove the conmented code?

File: include/AudioFileMP3.h [hunk 5]
```diff
@@ -58,7 +58,7 @@ class AudioFileMP3 : public AudioFileDevice
 	}
 
 protected:
-	void writeBuffer(const SampleFrame* /* _buf*/, const fpp_t /*_frames*/) override;
+	void writeBuffer(const SampleFrame* /* _buf*/, const size_t /*_frames*/) override;
```
Comment (Rossmaxx) [inline include/AudioFileMP3.h:61]: The other audio files doesn't seem to have the commented description of arguments.

File: plugins/AudioFileProcessor/AudioFileProcessor.cpp [hunk 6]
```diff
@@ -276,7 +276,7 @@ QString AudioFileProcessor::nodeName() const
 
 
 
-auto AudioFileProcessor::beatLen(NotePlayHandle* note) const -> int
+auto AudioFileProcessor::beatLen(NotePlayHandle* note) const -> size_t
```
Comment (Rossmaxx) [inline plugins/AudioFileProcessor/AudioFileProcessor.cpp:279]: Why do we use auto and trailing data type here? Isn't the old way better?

File: plugins/MultitapEcho/MultitapEcho.cpp [hunk 7]
```diff
@@ -118,8 +118,8 @@ bool MultitapEchoEffect::processAudioBuffer( SampleFrame* buf, const fpp_t frame
 	}
 	
 	// add dry buffer - never swap inputs for dry
-	m_buffer.writeAddingMultiplied( buf, 0, frames, dryGain );
-	
+	m_buffer.writeAddingMultiplied(buf, static_cast<size_t>(0), frames, dryGain);
```
Comment (Rossmaxx) [inline plugins/MultitapEcho/MultitapEcho.cpp:121]: Why do we need to do a static cast here? Wouldn't it be better to change the calling function, assuming all callers are changed.

**Review by Rossmaxx — COMMENTED** [review]
A few minor suggestions. The style changes seem to be too much so feel free to just resolve what I pointed out and neglect the rest 

Also, you added `mingw-std-threads` back in. Minor oversight.

File: include/AudioDevice.h (hunk 4 above)
Comment (Rossmaxx) [reply include/AudioDevice.h:96]: ie, the arguments

**Comment by sakertooth** [comment]
> Also, you added `mingw-std-threads` back in. Minor oversight.

Not an oversight, I just don't know how to fix it. This branch does not have `mingw-std-threads`, surely. It is up to date with `master` already.

**Comment by Rossmaxx** [comment]
Delete the submodule folder manually and commit

**Comment by sakertooth** [comment]
Oh nevermind you're totally right, I added it in somehow :thinking:.

**Comment by Rossmaxx** [comment]
That seems to be the reason for ci fail too

**Comment by sakertooth** [comment]
Also, I'm not going to worry about style for this PR. For these kind of PRs that span across many files, its far too time consuming to format each of the changes.

**Comment by Rossmaxx** [comment]
> Also, I'm not going to worry about style for this PR.

Alright then

**Comment by DomClark** [comment]
> There were discussions about using `std::size_t`, but I left it as `size_t` so that it can be adopted a bit easier in the code, and not end up in a situation where we mix up `std::size_t` and `size_t`.

They're the same type, just a different name, so there's no concern about mixing them up. I'm also not sure how using plain `size_t` makes it easier to adopt (besides being five characters shorter, but programming isn't about minimising the character count) - all it does is give us something we need to change yet again in the future, which sounds like more work to me.

A smaller change, simply redefining `fpp_t` and `f_cnt_t` would have been fine too, given that we want to replace most of these with `std::span` in the future.

**Comment by sakertooth** [comment]
> They're the same type, just a different name, so there's no concern about mixing them up. I'm also not sure how using plain size_t makes it easier to adopt (besides being five characters shorter, but programming isn't about minimising the character count) - all it does is give us something we need to change yet again in the future, which sounds like more work to me.

I figured that developers currently are more familiar with specifying  `size_t` rather than `std::size_t`, and catching the use of `size_t` instead of `std::size_t` in each new PR might be a lot of work. In that case, inevitably, the code has a high chance of having both `size_t` and `std::size_t`, and I like it more when a codebase is consistent in its choices and how its being read.

> A smaller change, simply redefining fpp_t and f_cnt_t would have been fine too, given that we want to replace most of these with std::span in the future.

Yeah, this is easier to do. I will do that instead and define them as `std::size_t`. :+1:

**Comment by sakertooth** [comment]
> A smaller change, simply redefining fpp_t and f_cnt_t would have been fine too, given that we want to replace most of these with std::span in the future.

I just realized we already have some code now that uses `size_t` instead of `fpp_t` or `f_cnt_t` for frame count, so replacing it throughout the codebase might actually be more ideal for consistency and readability, and then more ideal obviously would be to use `std::span` later. Since this PR has some specific changes to make the code compile though, replacing it throughout the codebase here would mask those specific changes as it would blow up the diff, which should be reviewed if possible.

**Comment by Rossmaxx** [comment]
You still didn't remove the submodule 😅

**Comment by sakertooth** [comment]
> You still didn't remove the submodule 😅

I did. It's not here anymore.

**Comment by sakertooth** [comment]
> Also, I'm not going to worry about style for this PR. For these kind of PRs that span across many files, its far too time consuming to format each of the changes.

Nevermind, since I'm not worrying about replacing `fpp_t` and `f_cnt_t` with `size_t` across the codebase, it's fine to consider style. These variables can slowly be phased out over time in other PRs.

File: src/core/Instrument.cpp [hunk 8]
```diff
@@ -185,7 +186,7 @@ void Instrument::applyRelease( SampleFrame* buf, const NotePlayHandle * _n )
 	const auto releaseFrames = desiredReleaseFrames();
 
 	const auto endFrame = _n->framesLeft();
-	const auto startFrame = std::max(0, endFrame - releaseFrames);
+	const auto startFrame = std::max(0, static_cast<int>(endFrame) - static_cast<int>(releaseFrames));
```
Comment (DomClark) [inline src/core/Instrument.cpp:189]: How about this?
```suggestion
	const auto startFrame = endFrame - std::min(endFrame, releaseFrames);
```
This avoids the unsigned subtraction problem while maintaining appropriate types.

File: plugins/Vibed/VibratingString.cpp [hunk 9]
```diff
@@ -99,7 +100,7 @@ void VibratingString::resample(const float* src, f_cnt_t srcFrames, f_cnt_t dstF
 	{
 		const float srcFrameFloat = frame * static_cast<float>(srcFrames) / dstFrames;
 		const float fracPos = srcFrameFloat - static_cast<f_cnt_t>(srcFrameFloat);
-		const f_cnt_t srcFrame = std::clamp(static_cast<f_cnt_t>(srcFrameFloat), 1, srcFrames - 3);
+		const f_cnt_t srcFrame = std::clamp<f_cnt_t>(static_cast<f_cnt_t>(srcFrameFloat), 1, srcFrames - 3);
```
Comment (DomClark) [inline plugins/Vibed/VibratingString.cpp:103]: I'd rather avoid specifying the template argument, and instead make sure we're passing values of the right type. This ensures the template always uses the appropriate type for the arguments, and avoids implicit conversions (in fact, if you're manually specifying the template argument to fix an error, you're almost certainly relying on such conversions).
```suggestion
		const f_cnt_t srcFrame = std::clamp(static_cast<f_cnt_t>(srcFrameFloat), f_cnt_t{1}, srcFrames - 3);
```
There are a couple of other cases to address involving `clamp`, `max`, etc.

**Review by DomClark — APPROVED** [review]

**Comment by zonkmachine** [comment]
@sakertooth I bisected master and these changes causes the issue where LB302 is silent when first launched. Just drag an LB302 into a new project and press the instrument keys. Silent. Add a note in the project and press play. Sound. Now you can press the gui keys and there will be sound from the LB302. Something isn't initialized properly.

Issue noticed by @bratpeki here: https://github.com/LMMS/lmms/pull/7484#issuecomment-2357035323

**Comment by sakertooth** [comment]
> @sakertooth I bisected master and these changes causes the issue where LB302 is silent when first launched. Just drag an LB302 into a new project and press the instrument keys. Silent. Add a note in the project and press play. Sound. Now you can press the gui keys and there will be sound from the LB302. Something isn't initialized properly.
> 
> 
> 
> Issue noticed by @bratpeki here: https://github.com/LMMS/lmms/pull/7484#issuecomment-2357035323

Thanks @zonkmachine for letting me know. I'll take a look today.

**Pull request merged**